File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()) return false;
        HashMap<Character, Integer> h1 = new HashMap<>();
        HashMap<Character, Integer> h2 = new HashMap<>();

        for(int i=0;i<s1.length();i++){
            h1.put(s1.charAt(i), h1.getOrDefault(s1.charAt(i),0)+1);
            h2.put(s2.charAt(i), h2.getOrDefault(s2.charAt(i),0)+1);
        }

        if(h1.equals(h2)){
            return true;
        }

        int left = 0;
        for(int right = s1.length(); right < s2.length(); right++){
            h2.put(s2.charAt(right), h2.getOrDefault(s2.charAt(right),0)+1);
            
            h2.put(s2.charAt(left), h2.getOrDefault(s2.charAt(left),0)-1);
            if(h2.get(s2.charAt(left)) == 0){
                h2.remove(s2.charAt(left));
            }
            left++;

            if(h1.equals(h2)){
                return true;
            }


        }

        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
            
        h1 = {}
        h2 = {}

        for i in range(len(s1)):
            if s1[i] not in h1:
                h1[s1[i]] = 1
            else:
                h1[s1[i]] += 1

            if s2[i] not in h2:
                h2[s2[i]] = 1
            else:
                h2[s2[i]] += 1

        if h1 == h2:
            return True

        left = 0
        for i in range(len(s1), len(s2)):
            if s2[i] in h2:
                h2[s2[i]] += 1
            else:
                h2[s2[i]] = 1
            
            if s2[left] in h2:
                h2[s2[left]] -= 1

            if h2[s2[left]] == 0:
               del h2[s2[left]]

            if h1 == h2:
                return True
            
            left += 1

        return False

        
